### PR TITLE
default to discovered server resources

### LINKED ISSUE #41353: Use of hardcoded "UserResources" breaks `kubectl get all` if server does not have at least one of the resources
**Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"6+", GitVersion:"v1.6.0-alpha.1.778+7585c46bc18913-dirty", GitCommit:"7585c46bc18913232a72d83e2f5a02a5a1f65c1b", GitTreeState:"dirty", BuildDate:"2017-02-13T18:47:18Z", GoVersion:"go1.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"", Minor:"", GitVersion:"v0.0.0-master+$Format:%h$", GitCommit:"$Format:%H$", GitTreeState:"not a git tree", BuildDate:"1970-01-01T00:00:00Z", GoVersion:"go1.7.3", Compiler:"gc", Platform:"linux/amd64"}

Using `kubectl` against an api server that does not have at least a [single resource in this list defined](
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/shortcut_restmapper.go#L105-L113) causes `kubectl get all` to fail with:

```
the server doesn't have a resource type "pods"
```

Instead of defaulting to the hardcoded list of resources, we could instead attempt to use the discovered resources from the server (if any were found). 

**How to reproduce it** (as minimally and precisely as possible):
`kubectl get all --server=<address of service catalog api server>`

cc @pmorie

### PR DESCRIPTION
**Pull request by juanvallejo**
Fixes #41353

When expanding "kubectl get all", this patch defaults the list of resources returned from shortcut_restmapper to those discovered on the server (if any). This allows "kubectl get all" to be successfully used against servers that do not have defined any of these hardcoded [UserResources](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/shortcut_restmapper.go#L105-L114) resources.

**Special notes for your reviewer**:
Relevant when using kubectl against the [service catalog](https://github.com/kubernetes-incubator/service-catalog) api-server

**Release note**:
```release-note
NONE
```

cc @pmorie @pwittrock @fabianofranz @kubernetes/sig-cli

### CODE REVIEW

**Comment by janetkuo** [comment]
@juanvallejo if no release note is required, put `NONE` in the release note block

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 1]
```diff
@@ -67,7 +69,19 @@ func (e ShortcutExpander) getAll() []schema.GroupResource {
 		}
 	}
 
-	return availableAll
+	// populate all "discovered" resources from the server
+	for availableResource := range availableResources {
+		r := availableResource.GroupResource()
```
Comment (k82cn) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: AFAIK, range slice will return index if only one lvalue, and int did not have `GroupResource` funcs. may be:

```
  for _, availableResource := range availableResources {
    ...
  }
```

**Review by k82cn — COMMENTED** [review]
Please add unit test for this case.

Just an overall review; I'll check its functional part later :).

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 2]
```diff
@@ -40,20 +40,22 @@ func NewShortcutExpander(delegate meta.RESTMapper, client discovery.DiscoveryInt
 	return ShortcutExpander{All: UserResources, RESTMapper: delegate, discoveryClient: client}
 }
 
-func (e ShortcutExpander) getAll() []schema.GroupResource {
+func (e ShortcutExpander) getAll() ([]schema.GroupResource, []schema.GroupResource) {
```
Comment (liggitt) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: this signature for a shortcut expander doesn't make sense

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 3]
```diff
@@ -67,7 +69,19 @@ func (e ShortcutExpander) getAll() []schema.GroupResource {
 		}
 	}
 
-	return availableAll
+	// populate all "discovered" resources from the server
+	for availableResource := range availableResources {
```
Comment (liggitt) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: not sure defaulting to all discovered resources is correct, given the [current definition of "all"](https://github.com/kubernetes/community/blob/master/contributors/devel/kubectl-conventions.md#rules-for-extending-special-resource-alias---all)

**Review by liggitt — CHANGES_REQUESTED** [review]

File: pkg/kubectl/cmd/util/shortcut_restmapper.go (hunk 1 above)
Comment (juanvallejo) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: @k82cn Thanks for the feedback, although I am using the index, in this case the map `availableResources` contains the actual groupVersionResource as the key, and an empty struct as the value

File: pkg/kubectl/cmd/util/shortcut_restmapper.go (hunk 2 above)
Comment (juanvallejo) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: hm, maybe it'd be better to check if any of the hardcoded `UserResources` matched any of the  available server resources here and if the final list is empty, return `availableResources` as the only slice?

File: pkg/kubectl/cmd/util/shortcut_restmapper.go (hunk 3 above)
Comment (juanvallejo) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: Hm, this just became a bit more tricky. Other than maybe specifying the contents of `UserResources` as part of a client config, I am not sure how else to handle cases where none of the resources in `UserResources` match any of the discovered ones.

Comment (juanvallejo) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: Added a second commit that filters out discovered resources that are not namespaced, however still need to filter out admin-level and potentially unrecoverable resources from the list

**Comment by fabianofranz** [comment]
Closing in favor of the one I'm about to open.

**Pull request closed without merging**